Add ERC: Subscription-Based Token (#60)
ethereum/ERCs: ERCS/erc-6932.md

title: Subscription-Based Token
description: ERC-20 extension providing access to a service or product that requires recurring payments.
author: 360 Core <hello@360coreinc.com>, Robin Rajput (@0xRobinR)
discussions-to: https://ethereum-magicians.org/t/erc-subscription-based-erc20-token/13964
status: Draft
type: Standards Track
category: ERC
created: 2023-04-25
requires: 20
---

## Abstract

The subscription-based [ERC-20](./eip-20.md) token extends the basic [ERC-20](./eip-20.md) token standard with a `subscribe` and `unsubscribe` function, which allow users to subscribe or unsubscribe from the subscription service. The `subscriptionFee` and `subscriptionFrequency` variables define the cost and frequency of the subscription. The `nextPaymentDate` mapping keeps track of the next payment date for each subscriber.

This EIP proposes the standardization of a subscription token on the Ethereum blockchain, enabling automatic periodic deductions from user balances as determined by the merchant subscriber. The Subscription Token aims to simplify and streamline subscription-based services on the Ethereum network, offering enhanced convenience and efficiency for users and merchants alike.

> We will be also adding `renewSubscription` functions to the token contract, that can be used by token holders to renew their subscription to a service or product that requires recurring payments in the form of the token.

Reviewer: You should write proposals as if they were already in the Final status. This line won't make sense once this proposal is Final.

## Specification

The Subscription Token ensures consistency and interoperability across different implementations. Key features include:

- Auto Deduction: Merchants, acting as subscribers, can set the subscription interval and associated payment amount for their services. This information is encoded within the Subscription Token contract, enabling automatic deductions from user balances at regular intervals without requiring manual intervention.

- Balance Check: Users can verify the remaining balance of their subscription tokens at any given time. This transparency empowers users to monitor their subscriptions and make informed decisions regarding their ongoing commitment to the service.

- Flexibility: The Subscription Token framework accommodates various subscription models, such as monthly, quarterly, or annual billing cycles. Additionally, merchants have the option to define trial periods, upgrade/downgrade plans, and cancellation policies, providing a versatile foundation for a wide range of subscription-based businesses.

> - Security: The Subscription Token employs established security measures, including the use of cryptographic signatures, to ensure the integrity and authenticity of subscription-related transactions. This protects both users and merchants from unauthorized access and potential malicious activities.

Reviewer: None of these lines specify requirements on implementations. They read more like Motivation because they explain the benefits of this standard.

> ### SubscriptionToken

Reviewer: Please put code snippets (like contract/function identifiers) in backticks (\`). This goes for all instances, not just this line.

```suggestion
### `SubscriptionToken`
```

#### subscribers

> returns the list of `addresses` subscribed to the subscription token contract.

Reviewer: ```suggestion
Returns the list of `addresses` subscribed to the subscription token contract.
```

#### subscriptionInfo

> metadata information of the subscription, like - `subscriptionID`, `subscriptionName`, `subscriptionDesc` and `subscriptionTandC`.

Reviewer: ```suggestion
Metadata information of the subscription, like - `subscriptionID`, `subscriptionName`, `subscriptionDesc` and `subscriptionTandC`.
```

#### subscriptionFee

> the subscription amount specified that will be deducted in `subscriptionFrequency` interval, when an address subscribes to the subscription token contract.

Reviewer: ```suggestion
The subscription amount specified that will be deducted in `subscriptionFrequency` interval, when an address subscribes to the subscription token contract.
```

Author: updated all

#### subscriptionFrequency

> frequency of subscription, interval at which the `subscriptionFee` will be charged. for example, every 1 day, 1 week or 1 month, denoted in seconds.

Reviewer: ```suggestion
Frequency of subscription, interval at which the `subscriptionFee` will be charged. for example, every 1 day, 1 week or 1 month, denoted in seconds.
```

#### subscribe

> method for subscribing an address to the subscription token contract.

Reviewer: ```suggestion
Method for subscribing an address to the subscription token contract.
```

#### unsubscribe

> revoke subscription from the subscription token contract, by the subscribed address.

Reviewer: ```suggestion
Revoke subscription from the subscription token contract, by the subscribed address.
```

  /// @return subscriptionID, subscriptionName, subscriptionDesc, subscriptionTandC
  function subscriptionInfo() external view returns ( uint, string memory, string memory, string memory );

  /// @dev subscribes to the subscription, can be payable
  function subscribe() external;

  /// @dev unsubscribe the subscription
  function unsubscribe() external;

  /// @dev view or pure can be used
  /// @return the subscription fee
  function subscriptionFee() external view returns (uint256);

  /// @dev view or pure can be used
  /// @return get the subscription frequency
  function subscriptionFrequency() external view returns (uint);
}
```

## Rationale

The subscription token contract inherits the fundamentals of subscription by deducting payments from subscribed addresses on a regular interval using mathematical formulas.

```
uint256 intervals = ( block.timestamp - info.start ) / info.frequency;
uint256 amount = info.amount * intervals;

uint256 localEffectiveBalance = effectiveBalance[account];

if ( (totalAmount + amount) > localEffectiveBalance ) {
    amount = localEffectiveBalance;
}

totalAmount += ( localEffectiveBalance - amount );
```

> here, the token balance of the address is calcuated using, the locked balances from ongoing subscripitons and the effective balance of the address (updates whenever a transfer is made).

Reviewer: ```suggestion
Here, the token balance of the address is calculated using the locked balances from ongoing subscriptions and the effective balance of the address (updates whenever a transfer is made).
```

## Specification

Below are the implementaions required by the standard:

### `SubscriptionToken`

#### `subscribers`

Returns the list of `addresses` subscribed to the subscription token contract.

#### `subscriptionInfo`

Metadata information of the subscription, like - `subscriptionID`, `subscriptionName`, `subscriptionDesc` and `subscriptionTandC`.

#### `subscriptionFee`

The subscription amount specified that will be deducted in `subscriptionFrequency` interval, when an address subscribes to the subscription token contract.

#### `subscriptionFrequency`

Frequency of subscription, interval at which the `subscriptionFee` will be charged. for example, every 1 day, 1 week or 1 month, denoted in seconds.

#### `subscribe`

Method for subscribing an address to the subscription token contract.

#### `unsubscribe`

Revoke subscription from the subscription token contract, by the subscribed address.

> ```solidity

Reviewer: This code block will appear under the `unsubscribe` heading in the Table of Contents. I'd recommend moving it directly under the `SubscriptionToken` heading so it makes more sense.
